File: src/main/java/space/maxkonkin/nasapicbot/web/NASAPicOnSpringBot.java
package space.maxkonkin.nasapicbot.web;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.FieldDefaults;
import org.springframework.beans.factory.annotation.Autowired;
import org.telegram.telegrambots.meta.api.methods.BotApiMethod;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.methods.updates.SetWebhook;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.starter.SpringWebhookBot;
import space.maxkonkin.nasapicbot.exception.UserNotFoundException;
import space.maxkonkin.nasapicbot.model.LangCode;
import space.maxkonkin.nasapicbot.model.NasaObject;
import space.maxkonkin.nasapicbot.model.User;
import space.maxkonkin.nasapicbot.service.NasaService;
import space.maxkonkin.nasapicbot.service.TranslateService;
import space.maxkonkin.nasapicbot.service.UserService;
import space.maxkonkin.nasapicbot.util.NasaObjectUtil;

import java.io.IOException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


@Getter
@Setter
@FieldDefaults(level = AccessLevel.PRIVATE)
public class NASAPicOnSpringBot extends SpringWebhookBot {
    public static final String HELP_TEXT = """
            Привет, я бот NASA! Я высылаю ссылки на картинки (или видео) с описанием по запросу. Введи команду:
            /give чтобы получить сегодняшнюю картинку;
            /random чтобы получить случайную картинку.
            Либо введи дату в формате <b>YYYY-MM-DD</b> и я пришлю ссылку на картинку с описанием, опубликованную в тот день.
            Дата должна быть не раньше 1995-06-20!
            Напоминаю, что картинки на сайте NASA обновляются раз в сутки""";
    String botPath;
    String botUsername;
    String errorText;
    Boolean withTranslate;

    private final NasaService nasaService;

    private final TranslateService translateService;

    private final UserService userService;

    public NASAPicOnSpringBot(SetWebhook setWebhook, String botToken, NasaService nasaService, TranslateService translateService, UserService userService) {
        super(setWebhook, botToken);
        this.nasaService = nasaService;
        this.translateService = translateService;
        this.userService = userService;
    }

    @Override
    public BotApiMethod<?> onWebhookUpdateReceived(Update update) {
        try {
            return handleUpdate(update);
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }
        return null;
    }

    public SendMessage handleUpdate(Update update) throws IOException {
        if (!update.hasCallbackQuery()) {
            if (update.hasMessage()) {
                Message message = update.getMessage();
                long chat_id = message.getChatId();

                var tgUser = message.getFrom();
                var newUser = new User(chat_id, tgUser.getUserName(), false, LangCode.EN);
                userService.saveNew(newUser);
                User user = userService.getById(chat_id).orElseThrow(() ->
                        new UserNotFoundException("user with chat_id=" + chat_id + " not found"));

                String text = message.getText();
                final String regex = "\\d{4}-\\d{2}-\\d{2}";
                final Pattern pattern = Pattern.compile(regex);
                assert text != null;
                final Matcher matcher = pattern.matcher(text);
                if (matcher.find()) {
                    String fromRegex = matcher.group(0);
                    try {
                        LocalDate date = LocalDate.parse(fromRegex, DateTimeFormatter.ISO_LOCAL_DATE);
                        if (!date.isBefore(LocalDate.of(1995, 6, 20)) &&
                                !date.isAfter(LocalDate.now())) {
                            return givePostedOnDatePicture(date, user);
                        } else {
                            return sendMessage("Введённая дата должна быть не раньше 1995-06-20 и не позже сегодняшней даты", chat_id);
                        }
                    } catch (Exception e) {
                        System.err.println("Parsing error! " + e.getMessage());
                        return sendMessage("Неверный формат даты.\nВведите дату в формате <b>YYYY-MM-DD</b>", chat_id);
                    }
                } else {
                    switch (text) {
                        case "/start", "/help" -> {
                            return sendMessage(HELP_TEXT, chat_id);
                        }
                        case "/give" -> {
                            return giveTodayPicture(user);
                        }
                        case "/random" -> {
                            return giveRandomPicture(user);
                        }
                        default -> {
                            return sendMessage(errorText, chat_id);
                        }
                    }
                }
            }
        }
        return null;
    }

    private SendMessage giveRandomPicture(User user) throws IOException {
        NasaObject random = nasaService.getRandom();
        assert random != null;
        if (withTranslate && user.getTranslateLangCode() != LangCode.EN) {
            return sendTranslatedAndFormattedMessage(random, user);
        } else {
            return sendFormattedMessage(random, user.getChatId());
        }
    }

    public SendMessage giveTodayPicture(User user) throws IOException {
        NasaObject today = nasaService.getToday();
        assert today != null;
        if (withTranslate && user.getTranslateLangCode() != LangCode.EN) {
            return sendTranslatedAndFormattedMessage(today, user);
        } else {
            return sendFormattedMessage(today, user.getChatId());
        }
    }

    private SendMessage givePostedOnDatePicture(LocalDate date, User user) throws IOException {
        NasaObject onDate = nasaService.getOnDate(date);
        assert onDate != null;
        if (withTranslate && user.getTranslateLangCode() != LangCode.EN) {
            return sendTranslatedAndFormattedMessage(onDate, user);
        } else {
            return sendFormattedMessage(onDate, user.getChatId());
        }
    }

    private SendMessage sendFormattedMessage(NasaObject nasaObject, long chat_id) {
        return sendMessage(NasaObjectUtil.getFormattedMessage(nasaObject), chat_id);
    }

    private SendMessage sendTranslatedAndFormattedMessage(NasaObject nasaObject, User user) throws IOException {
        NasaObject translated = translateService.translateTitleAndExplanation(nasaObject, user.getTranslateLangCode());
        return sendMessage(NasaObjectUtil.getFormattedMessage(translated), user.getChatId());
    }

    private SendMessage sendMessage(String messageText, long chat_id) {
        SendMessage message = new SendMessage();
        message.setChatId(chat_id);
        message.setText(messageText);
        message.enableHtml(true);
        return message;
    }
}

File: src/main/kotlin/space/maxkonkin/nasapicbot/web/NASAPicOnSpringBot.kt
package space.maxkonkin.nasapicbot.web

import org.telegram.telegrambots.meta.api.methods.botapimethods.BotApiMethod
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.webhook.TelegramWebhookBot
import space.maxkonkin.nasapicbot.exception.UserNotFoundException
import space.maxkonkin.nasapicbot.model.BotCommand
import space.maxkonkin.nasapicbot.model.LangCode
import space.maxkonkin.nasapicbot.model.ScheduleState
import space.maxkonkin.nasapicbot.model.Token
import space.maxkonkin.nasapicbot.model.User
import space.maxkonkin.nasapicbot.service.MessageService
import space.maxkonkin.nasapicbot.service.NasaService
import space.maxkonkin.nasapicbot.service.UserService
import space.maxkonkin.nasapicbot.to.NasaTo
import space.maxkonkin.nasapicbot.util.getFormattedMessage

class NASAPicOnSpringBot(
    private val nasaService: NasaService,
    private val userService: UserService,
    private val messageService: MessageService,
    private val botPath: String,
    private val botName: String,
    private val setWebhook: Runnable? = null,
    private val deleteWebhook: Runnable? = null
) : TelegramWebhookBot {

    override fun runDeleteWebhook() {
        deleteWebhook?.run()
    }

    override fun runSetWebhook() {
        setWebhook?.run()
    }

    override fun consumeUpdate(update: Update): BotApiMethod<*>? {
        return handleUpdate(update, null)
    }

    override fun getBotPath(): String? {
        return botPath
    }

    fun handleUpdate(update: Update, token: Token?): SendMessage? {
        if (update.hasCallbackQuery()) return null
        if (!update.hasMessage()) return null

        val message = update.message
        val chatId = message.chatId
        val newUser = User(chatId, message.from.userName, ScheduleState.NONE, LangCode.EN)
        userService.saveNew(newUser)
        val user = userService.getById(chatId)
            ?: throw UserNotFoundException("user with chat_id=$chatId not found")

        return when (val cmd = BotCommand.parse(checkNotNull(message.text), botName, message.isReply)) {
            is BotCommand.Start, BotCommand.Help ->
                sendMessage(messageService.getMessage("help.text", user.locale()), chatId)

            is BotCommand.Today -> giveTodayPicture(user)
            is BotCommand.Random -> giveRandomPicture(user)
            is BotCommand.Schedule -> toggleSchedule(user, token)
            is BotCommand.Translate -> toggleTranslate(user)
            is BotCommand.OnDate -> givePostedOnDatePicture(cmd.date, user)
            is BotCommand.InvalidDateRange ->
                sendMessage(messageService.getMessage("error.invalid_date_range", user.locale()), chatId)

            is BotCommand.InvalidDateFormat ->
                sendMessage(messageService.getMessage("error.invalid_date_format", user.locale()), chatId)

            is BotCommand.Unknown ->
                sendMessage(messageService.getMessage("error.unsupported_command", user.locale()), chatId)

            is BotCommand.Reply -> null
        }
    }

    private fun giveRandomPicture(user: User): SendMessage {
        val random = nasaService.getRandom(user)
        return sendFormattedMessage(user, requireNotNull(random) { "Unable to send message" }, user.chatId)
    }

    fun giveTodayPicture(user: User): SendMessage {
        val today = nasaService.getToday(user)
        return sendFormattedMessage(user, requireNotNull(today) { "Unable to send message" }, user.chatId)
    }

    private fun givePostedOnDatePicture(date: java.time.LocalDate, user: User): SendMessage {
        val onDate = nasaService.getOnDate(date, user)
        return sendFormattedMessage(user, requireNotNull(onDate) { "Unable to send message" }, user.chatId)
    }

    private fun toggleSchedule(user: User, token: Token?): SendMessage {
        val newState = userService.toggleSchedule(user, token)
        return sendMessage(messageService.getScheduleMessage(newState, user.locale()), user.chatId)
    }

    private fun toggleTranslate(user: User): SendMessage {
        val updatedUser = if (user.translateLangCode != LangCode.EN) {
            user.copy(translateLangCode = LangCode.EN)
        } else {
            user.copy(translateLangCode = LangCode.RU)
        }
        userService.update(updatedUser)
        val isEnabled = updatedUser.translateLangCode != LangCode.EN
        return sendMessage(messageService.getTranslationMessage(isEnabled, updatedUser.locale()), updatedUser.chatId)
    }

    private fun sendFormattedMessage(user: User, nasaTo: NasaTo, chatId: Long): SendMessage {
        return sendMessage(
            getFormattedMessage(nasaTo, messageService.getMessage("message.posted_on", user.locale())),
            chatId
        )
    }

    private fun sendMessage(messageText: String, chatId: Long): SendMessage {
        return SendMessage.builder()
            .parseMode("HTML")
            .chatId(chatId)
            .text(messageText)
            .build()
    }

    private fun User.locale(): String = translateLangCode.code
}